firedrake/linear_solver.py

```diff
@@ -116,36 +118,27 @@ def _rhs(self):
 
         u = function.Function(self.trial_space)
         b = cofunction.Cofunction(self.test_space.dual())
-        expr = -action(self.A.a, u)
-        return u, get_assembler(expr).assemble, b
+        expr = b - action(self.A.a, u)
+        return u, get_assembler(expr, bcs=self.A.bcs, zero_bc_nodes=False).assemble, b
 
     def _lifted(self, b):
         u, update, blift = self._rhs
         u.dat.zero()
         for bc in self.A.bcs:
             bc.apply(u)
+        blift.assign(b)
         update(tensor=blift)
```

Comment:
`blift` (`b` in `def _rhs()`) is now input as well as output? Am I missing something?

Reply:
Is this a memory leak? https://github.com/firedrakeproject/firedrake/blob/ecac12bc9763f0e37e62bf5e32d5bff17d6067cd/firedrake/assemble.py#L482

Reply:
I have changed the expression for `b - A*x` to the original - A*x but now the update is done by first copying the incoming residual into the output `tensor` and setting `needs_zeroing=False`. This avoids the creation of a new `Cofunction` for the `maxpy`.

Reply:
I think `update(tensor=blift)` pollutes BC node values of `blift` (by `-A*u_bc`), so we still need to explicitly set `bc.apply(blift)`, no?